### ISSUE TITLE
Ghost atoms with custom basis are not marked as ghost

### ISSUE THREAD

**Issue by susilehtola** [pyscf/pyscf#548]
Calculations crash for heavy atoms
```
  File "/wrk/users/jzlehtol/pyscf/pyscf/lib/misc.py", line 601, in aliased_fn
    return getattr(self, fname)(*args, **kwargs)
  File "/wrk/users/jzlehtol/pyscf/pyscf/scf/hf.py", line 1638, in scf
    conv_check=self.conv_check, **kwargs)
  File "/wrk/users/jzlehtol/pyscf/pyscf/scf/hf.py", line 122, in kernel
    dm = mf.get_init_guess(mol, mf.init_guess)
  File "/wrk/users/jzlehtol/pyscf/pyscf/scf/hf.py", line 1571, in get_init_guess
    dm = self.init_guess_by_atom(mol)
  File "/wrk/users/jzlehtol/pyscf/pyscf/scf/hf.py", line 1527, in init_guess_by_atom
    return init_guess_by_atom(mol)
  File "/wrk/users/jzlehtol/pyscf/pyscf/scf/hf.py", line 462, in init_guess_by_atom
    atm_scf = atom_hf.get_atm_nrhf(mol)
  File "/wrk/users/jzlehtol/pyscf/pyscf/scf/atom_hf.py", line 52, in get_atm_nrhf
    atm_hf.run()
  File "/wrk/users/jzlehtol/pyscf/pyscf/lib/misc.py", line 482, in run
    self.kernel(*args)
  File "/wrk/users/jzlehtol/pyscf/pyscf/lib/misc.py", line 601, in aliased_fn
    return getattr(self, fname)(*args, **kwargs)
  File "/wrk/users/jzlehtol/pyscf/pyscf/scf/atom_hf.py", line 139, in scf
    return hf.RHF.scf(self, *args, **kwargs)
  File "/wrk/users/jzlehtol/pyscf/pyscf/scf/hf.py", line 1638, in scf
    conv_check=self.conv_check, **kwargs)
  File "/wrk/users/jzlehtol/pyscf/pyscf/scf/hf.py", line 122, in kernel
    dm = mf.get_init_guess(mol, mf.init_guess)
  File "/wrk/users/jzlehtol/pyscf/pyscf/scf/hf.py", line 1580, in get_init_guess
    dm = self.init_guess_by_minao(mol)
  File "/wrk/users/jzlehtol/pyscf/pyscf/scf/hf.py", line 1521, in init_guess_by_minao
    return init_guess_by_minao(mol)
  File "/wrk/users/jzlehtol/pyscf/pyscf/scf/hf.py", line 419, in init_guess_by_minao
    occ_add, basis_add = minao_basis(symb, nelec_ecp)
  File "/wrk/users/jzlehtol/pyscf/pyscf/scf/hf.py", line 369, in minao_basis
    for b in basis_add[l][1:]])
IndexError: list index out of range
```
The problem is that the minao basis set only covers a small fraction of the periodic table.

The problem was fixed for the atom guess in 697f5e871e901795636397cc465519e72f2e8a6d, and broken again by 6ba5f9b1021bfe31397930a3293c5e091c240f7b.

**Comment by susilehtola** [comment]
It looks like the main issue I am facing with is that the guess isn't handling ghost atoms properly. Here's a reduced test case
```
#!/usr/bin/python3                                                                                                                                                                            

from pyscf import gto
from pyscf import dft, scf
import numpy
import os

mol = gto.M(
        verbose = 4,
        symmetry = 0,
        spin = 0,
        atom = '''                                                                                                                                                                            
Ba 0.0 0.0 -1.5                                                                                                                                                                               
ghost_Cn 0.0 0.0 0.000000000000000e+00                                                                                                                                                        
Ba 0.0 0.0 1.5                                                                                                                                                                                
''',
        basis = {
                    'Ba' : gto.basis.load('ugbs.nw', 'Ba'),
                    'Cn' : gto.basis.load('ugbs.nw', 'Cn')
                }
    )

method = scf.RHF(mol)
method = method.x2c1e()
method.kernel()
```

**Comment by susilehtola** [comment]
Indeed, the problem with the atom guess is that https://github.com/pyscf/pyscf/blob/fe0416a004317b9b1aa6125e3adce287cc2fe19f/pyscf/scf/atom_hf.py#L43 is not catching the ghost Cn atom.

For the minao `No basis found` problem, someone who knows the code should comment.

**Comment by fishjojo** [comment]
Ghost atoms are not treated in the MINAO basis for initial guesses.
See https://github.com/pyscf/pyscf/blob/fe0416a004317b9b1aa6125e3adce287cc2fe19f/pyscf/scf/hf.py#L354-L355
It shouldn't affect the result much though.

**Comment by susilehtola** [comment]
> Ghost atoms are not treated in the MINAO basis for initial guesses.
> It shouldn't affect the result much though.

Exactly - nothing should be done for ghost atoms in the initial guess!

**Comment by susilehtola** [comment]
Maybe the problem is the same in `minao` as in `atom`: the code flies through the ghost atom check. I'm guessing the `No basis found` error is generated three lines after at https://github.com/pyscf/pyscf/blob/fe0416a004317b9b1aa6125e3adce287cc2fe19f/pyscf/scf/hf.py#L358

**Comment by sunqm** [comment]
Problem was fixed in ff236bcb1bc77cfcc0103bc7861763f7696a7c40

**Comment by susilehtola** [comment]
@sunqm thanks, but I think the issue is deeper: the ghost atom checks are falling through!

**Comment by susilehtola** [comment]
`minao` is still printing the error `No basis found for atom GHOST-Cn` which should have been caught by https://github.com/pyscf/pyscf/blob/fe0416a004317b9b1aa6125e3adce287cc2fe19f/pyscf/scf/hf.py#L354-L355

Similarly, atomic RHF is running for `Cn` even though it's a ghost atom and should have been trapped by
https://github.com/pyscf/pyscf/blob/2fce03ce7bc67f974467dae3f347f2d72169cd7c/pyscf/scf/atom_hf.py#L43

The remaining problem is that the ghost atoms aren't marked as ghost if a custom basis set is used as in the example above.

**Comment by sunqm** [comment]
Which git stash is your version at?  With the updates in ff236bcb1bc77cfcc0103bc7861763f7696a7c40 and newer versions, I couldn't reproduce the error `No basis found for atom GHOST-Cn`

**Comment by susilehtola** [comment]
@sunqm I just merged master and recompiled the library too.
```
Date: Sun Mar 22 13:23:08 2020
PySCF version 1.7.1
PySCF path  /home/work/pyscf.susi/pyscf
GIT ORIG_HEAD 48fcf5a5cb182e8b6f9c924b0b96fcca9c3ae1b3
GIT HEAD      ref: refs/heads/master
GIT master branch  8015c2dbdf242200368e8b29b16a784bd8ec43a8
```
I still get the same results: the ghost atom checks fall through.

To get the basis not found error you need to change the initial guess to `minao`. 

As for `atom`, to see it is building a density matrix for the `Cn` ghost atom you can e.g. put a debug statement like `print('Running NRSRHF for {}, {}'.format(a,b))` at https://github.com/pyscf/pyscf/blob/8015c2dbdf242200368e8b29b16a784bd8ec43a8/pyscf/scf/atom_hf.py#L49

**Comment by btglau** [comment]
I think my experience is not relevant, but I was also tracking down an error of the type "No basis found for atom [...]", when calculating systems with ghost atoms using pyscf basis sets. For me, the error was printed in the invocation here:

https://github.com/pyscf/pyscf/blob/e498c0a7e9c0c669b65dd8ab54a3ab8049becfb6/pyscf/scf/hf.py#L438

Which I believe is building a new mol object without any of the ghost atoms. Therefore, in the make_env method, no basis is found for the atoms.

The only place I could find with the text "No basis found" occurs here:
https://github.com/pyscf/pyscf/blob/e498c0a7e9c0c669b65dd8ab54a3ab8049becfb6/pyscf/gto/mole.py#L840

**Comment by susilehtola** [comment]
But @btglau's issue is only affecting `minao`; a similar issue also exists in `atom`.

**Comment by sunqm** [comment]
Basis sanity check was done when parsing the basis files or basis strings. The check in gto/mole.py L840 is not needed any more. The inaccurate error message was removed.

**Comment by susilehtola** [comment]
@sunqm I repeat, @btglau's issue is unrelated to this report.

**Comment by sunqm** [comment]
Problem was fixed in commit ce35a520e010b069439a0ffaf2c5f4fd69561a53 . Ghost atoms are excluded from atomic HF calculations

**Comment by susilehtola** [comment]
Thanks @sunqm, I can confirm the issues are solved!
